cooked-validators/src/Cooked/MockChain/BlockChain.hs

```diff
@@ -13,49 +15,95 @@
 import Cooked.Skeleton
 import Data.Kind
 import Data.Maybe
-import qualified Ledger as Pl
-import qualified Ledger.Tx.CardanoAPI as Pl
+import qualified Ledger.Index as Ledger
+import qualified Ledger.Slot as Ledger
+import qualified Ledger.Tx as Ledger
+import qualified Ledger.Tx.CardanoAPI as Ledger
 import ListT
 import Optics.Core
+import qualified Plutus.Script.Utils.Scripts as Pl
 import qualified Plutus.V2.Ledger.Api as PV2
-import qualified PlutusTx as Pl
+import qualified Plutus.V2.Ledger.Api as Pl
 import Prettyprinter (Doc)
 
 -- * BlockChain Monad
 
-class (MonadFail m) => MonadBlockChainWithoutValidation m where
-  -- | Returns a list of all currently known outputs
-  allUtxos :: m [(Pl.TxOutRef, PV2.TxOut)]
+-- | The errors that can be produced by the 'MockChainT' monad
+data MockChainError where
+  MCEValidationError :: Ledger.ValidationErrorInPhase -> MockChainError
+  MCEUnbalanceable :: String -> BalanceStage -> TxSkel -> MockChainError
+  MCENoSuitableCollateral :: MockChainError
+  MCEGenerationError :: GenerateTxError -> MockChainError
+  MCECalcFee :: MockChainError -> MockChainError
+  MCEUnknownOutRefError :: String -> PV2.TxOutRef -> MockChainError
+  MCEUnknownValidator :: String -> PV2.ValidatorHash -> MockChainError
+  MCEUnknownDatum :: String -> PV2.DatumHash -> MockChainError
+  FailWith :: String -> MockChainError
+  OtherMockChainError :: (Show err, Eq err) => err -> MockChainError
+
+deriving instance Show MockChainError
+
+instance Eq MockChainError where
+  (==) = undefined
+
+-- | Describes us which stage of the balancing process are we at. This is needed
+--  to distinguish the successive calls to balancing while computing fees from
+--  the final call to balancing
+data BalanceStage
+  = BalCalcFee
+  | BalFinalizing
+  deriving (Show, Eq)
+
+class (MonadFail m, MonadError MockChainError m) => MonadBlockChainBalancing m where
+  -- | Returns the paramters of the chain.
+  getParams :: m Emulator.Params
+
+  -- | Return a list of all UTxOs at a certain address.
+  utxosAtLedger :: PV2.Address -> m [(PV2.TxOutRef, Ledger.TxOut)]
 
   -- | Returns the datum with the given hash, along with its pretty-printed
   -- representation, or 'Nothing' if there is none
-  datumFromHash :: Pl.DatumHash -> m (Maybe (Pl.Datum, Doc ()))
+  datumFromHash :: PV2.DatumHash -> m (Maybe (PV2.Datum, Doc ()))
+
+  -- | Returns the full validator corresponding to hash, if that validator is
+  -- currently the owner of some UTxO. (If it is not, there's no guarantee that
+  -- this will return @Just@.)
+  validatorFromHash :: PV2.ValidatorHash -> m (Maybe (Pl.Versioned PV2.Validator))
 
   -- | Returns an output given a reference to it
-  txOutByRef :: Pl.TxOutRef -> m (Maybe PV2.TxOut)
+  txOutByRefLedger :: PV2.TxOutRef -> m (Maybe Ledger.TxOut)
+
+class MonadBlockChainBalancing m => MonadBlockChainWithoutValidation m where
+  -- | Returns a list of all currently known outputs.
+  allUtxosLedger :: m [(PV2.TxOutRef, Ledger.TxOut)]
```

Comment:
It's a tad surprising to see `utxosAtLedger` and `allUtxosLedger` separated. But not a big deal!